### PR TITLE
Use only webpack assets.

### PR DESCRIPTION
**Pull request by tofumatt**
## Summary

Addresses issue #993.

## Relevant technical choices

This keeps `React` as a provided global throughout the project. I'm not usually a fan of this approach, but it makes this already large diff a lot easier to digest, and we can always look at changing that later.

This PR means we're entirely separate from the provided WordPress globals for `wp.` and React. This means we can use later versions of React, etc.

Technically it means we can opt-out of the `@wordpress/element` dependency entirely, as that largely exists to provide a consistent API to React for Gutenberg plugins. That said, there's not a huge problem keeping it around.

I tried to get bundlesize to check on the vendor sizes as well, but it would error every time I tried to check `dist/assets/js/vendor*.js` 🤷‍♂

* Internal `withFilters` HOC has been replaced with the core package again after confirming that the production build does not contain any added translation strings from that package (which was the reason for removing initially)
* Moved VRT to its own CI job
    - This makes the build more efficient by not requiring `composer install` for JS Tests
    - Also clearer as to why JS tests failed
* Storybook no longer requires a full build just for static assets and CSS
    - Static assets (svg + images) are built separately and storybook webpack config handles `scss` files instead of css  
    - This makes Storybook more viable for component development as now JS and styles are hot reloadable
* Test JS entrypoints are no longer built during the production build.
  - This reduces the ZIP file size as it looks like they were never excluded from the build.
  - This reduces build times for production builds as it doesn't need to build the tests entrypoint
  - This meant a new `npm` task: `build:test` which does a production build but adds test code. I did this over using the development mode to ensure the E2E tests are run against production code.
* All global variables have been removed. React is still available globally via `WebpackProvide`, which means we don't need to `import React from 'react';` to write JSX. This is a nice quality-of-life improvement, especially when we import other things from React via `@wordpress/element`.

We should consider:

* Removing the E2E job for WP 4.9+Gutenberg, now that there are no longer any shared dependencies.

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Comment by aaemnnosttv** [comment]
@googlebot I consent.

File: includes/Core/Assets/Assets.php [hunk 1]
```diff
@@ -894,22 +894,49 @@ private function get_external_assets() {
...
 			// )
 			// ),
-			// new Script(
-			// 'wp-api-fetch',
-			// array(
-			// 'src'          => $base_url . 'js/externals/apiFetch.js',
-			// 'dependencies' => array( 'wp-i18n', 'wp-polyfill', 'wp-url' ),
-			// 'version'      => '3.6.4',
-			// 'fallback'     => true,
+			new Script('e2e-utilities', array(
```
Comment (aaemnnosttv) [inline includes/Core/Assets/Assets.php]: This shouldn't be registered in the main plugin. It should go in an mu-plugin which is only used in e2e tests. See https://github.com/google/site-kit-wp/tree/develop/tests/e2e/mu-plugins

Comment (tofumatt) [reply includes/Core/Assets/Assets.php]: Indeed! I'm in a bit of refactoring hell at the moment so this stuff isn't final, but I'll move it soon just for tidiness. 😄

File: includes/Core/Assets/Assets.php [hunk 2]
```diff
@@ -727,193 +749,4 @@ private function run_before_print_callbacks( WP_Dependencies $dependencies, arra
...
-					'src'          => $base_url . 'js/externals/wp-polyfill.js',
-					'version'      => '7.4.0',
-					'fallback'     => true,
-					'before_print' => function( $handle ) use ( $base_url ) {
-						$inline_polyfill_tests = array(
-							'\'fetch\' in window'                                    => $base_url . 'js/externals/wp-polyfill-fetch.js', // phpcs:ignore WordPress.Arrays.MultipleStatementAlignment
-							'document.contains'                                      => $base_url . 'js/externals/wp-polyfill-node-contains.js', // phpcs:ignore WordPress.Arrays.MultipleStatementAlignment
-							'window.FormData && window.FormData.prototype.keys'      => $base_url . 'js/externals/wp-polyfill-formdata.js', // phpcs:ignore WordPress.Arrays.MultipleStatementAlignment
-							'Element.prototype.matches && Element.prototype.closest' => $base_url . 'js/externals/wp-polyfill-element-closest.js', // phpcs:ignore WordPress.Arrays.MultipleStatementAlignment
-							'typeof URL === \'function\''                            => $base_url . 'js/externals/wp-polyfill-url.js', // phpcs:ignore WordPress.Arrays.MultipleStatementAlignment
```
Comment (aaemnnosttv) [inline includes/Core/Assets/Assets.php]: Are we sure that all of the polyfills we are no longer bundling aren't needed? Perhaps we need to re-add some to our `package.json`? Not just these checks specifically, but any of the polyfills from externals.

File: package.json [hunk 3]
```diff
@@ -43,11 +43,11 @@
...
       "path": "./dist/*.js",
-      "maxSize": "170 kB"
+      "maxSize": "270 kB"
     },
     {
-      "path": "./dist/assets/js/*.js",
-      "maxSize": "170 kB"
+      "path": "./dist/assets/js/googlesitekit-*.js",
+      "maxSize": "270 kB"
     }
```
Comment (aaemnnosttv) [inline package.json]: These were more important in the past I believe when the file size of the distributable was a limiting factor for some. Obviously we don't want to ship 50MB of js, but are bundle size tests still necessary?

File: assets/js/components/data/index.js [hunk 4]
```diff
@@ -204,7 +204,7 @@ const dataAPI = {
 		} ).then( ( results ) => {
 			each( results, ( result, key ) => {
 				if ( ! keyIndexesMap[ key ] ) {
-					console.error( 'data_error', 'unknown response key ' + key ); // eslint-disable-line no-console
+					console.error( 'data_error', 'unknown response key ' + key, addQueryArgs( '/google-site-kit/v1/data/', { datacache: datacache || undefined } ) ); // eslint-disable-line no-console
```
Comment (aaemnnosttv) [inline assets/js/components/data/index.js]: Why add the URL + datacache here? Datacache is only used in VRT.

File: webpack.config.js [hunk 5]
```diff
@@ -29,67 +29,7 @@ const glob = require( 'glob' );
...
-
-const externalEntry = {};
-externalPackages.forEach( ( packageName ) => {
-	const name = camelCaseDash( packageName );
-	externalEntry[ name ] = `./node_modules/@wordpress/${ packageName }`;
-} );
+const { ProvidePlugin } = require( 'webpack' );
 
 // This External Libraries will not part of wp object. Most of this is for Polyfill.
 const externalLibrary = {
```
Comment (aaemnnosttv) [inline webpack.config.js]: These shouldn't need to be built as external scripts anymore. Can we load the polyfills for all/some entrypoints using a webpack plugin perhaps?

File: tests/e2e/specs/modules/tagmanager/setup.test.js [hunk 6]
```diff
@@ -60,7 +60,7 @@ describe( 'Tag Manager module setup', () => {
 		// Intercept the call to window.open and call our API to simulate a created account.
 		await page.evaluate( () => {
 			window.open = () => {
-				window.wp.apiFetch( {
+				window._apiFetch( {
```
Comment (aaemnnosttv) [inline tests/e2e/specs/modules/tagmanager/setup.test.js]:
```suggestion
				window._e2eApiFetch( {
```

File: tests/e2e/specs/modules/analytics/setup-no-account-no-tag.test.js [hunk 7]
```diff
@@ -66,7 +66,7 @@ describe( 'setting up the Analytics module with no existing account and no exist
 		// Intercept the call to window.open and call our API to simulate a created account.
 		await page.evaluate( () => {
 			window.open = () => {
-				window.wp.apiFetch( {
+				window._apiFetch( {
```
Comment (aaemnnosttv) [inline tests/e2e/specs/modules/analytics/setup-no-account-no-tag.test.js]:
```suggestion
				window._e2eApiFetch( {
```

File: assets/js/api-fetch-shim.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
+/**
+ * WordPress dependencies
+ */
+import apiFetch from '@wordpress/api-fetch__non-shim';
+
+if ( global._apiFetch === undefined ) {
```
Comment (aaemnnosttv) [inline assets/js/api-fetch-shim.js]: Is using a global here necessary? We should only need to expose this via a global for e2e (which I see we have added as well). Would be great if these two could share the same code/global.

File: webpack.config.js [hunk 9]
```diff
@@ -188,13 +140,16 @@ module.exports = ( env, argv ) => {
...
 			resolve,
 		},
 
 		// Build the test files.
 		{
-			entry: { 'googlesitekit-tests': './assets/js/googlesitekit-tests.js' },
+			entry: {
+				'googlesitekit-tests': './assets/js/googlesitekit-tests.js',
+				'e2e-utilities': './tests/e2e/e2e-utilities.js',
+			},
```
Comment (aaemnnosttv) [inline webpack.config.js]: These are not being excluded from the release zip

<details>
<summary>Dist JS files</summary>

```
        0  01-27-2020 22:17   google-site-kit/dist/assets/js/
     1844  01-27-2020 22:17   google-site-kit/dist/assets/js/ads.js
   973157  01-27-2020 22:17   google-site-kit/dist/assets/js/allmodules.js
    44280  01-27-2020 22:17   google-site-kit/dist/assets/js/chunk-googlesitekit-adminbar-05e75b29d0d33cbe4709.js
    32177  01-27-2020 22:17   google-site-kit/dist/assets/js/chunk-googlesitekit-setup-wizard-8380113159d8cc93ff03.js
    23232  01-27-2020 22:17   google-site-kit/dist/assets/js/chunk-googlesitekit-setup-wizard-proxy-9e7a8c523a2f817a665f.js
    17925  01-27-2020 22:17   google-site-kit/dist/assets/js/chunk-googlesitekit-setup-wizard~chunk-googlesitekit-setup-wrapper-734b48a8cb72b8bd8603.js
     4473  01-27-2020 22:17   google-site-kit/dist/assets/js/chunk-googlesitekit-setup-wrapper-41213a0f28b700d924d7.js
    38127  01-27-2020 22:17   google-site-kit/dist/assets/js/e2e-utilities.js
        0  01-27-2020 22:17   google-site-kit/dist/assets/js/externals/
   606422  01-27-2020 22:17   google-site-kit/dist/assets/js/googlesitekit-activation.js
    13509  01-27-2020 22:17   google-site-kit/dist/assets/js/googlesitekit-admin.js
    15874  01-27-2020 22:17   google-site-kit/dist/assets/js/googlesitekit-adminbar-loader.js
   590003  01-27-2020 22:17   google-site-kit/dist/assets/js/googlesitekit-api.js
   627705  01-27-2020 22:17   google-site-kit/dist/assets/js/googlesitekit-dashboard-details.js
   630148  01-27-2020 22:17   google-site-kit/dist/assets/js/googlesitekit-dashboard-splash.js
   644158  01-27-2020 22:17   google-site-kit/dist/assets/js/googlesitekit-dashboard.js
   601452  01-27-2020 22:17   google-site-kit/dist/assets/js/googlesitekit-module.js
   727637  01-27-2020 22:17   google-site-kit/dist/assets/js/googlesitekit-settings.js
   554202  01-27-2020 22:17   google-site-kit/dist/assets/js/googlesitekit-tests.js
   589909  01-27-2020 22:17   google-site-kit/dist/assets/js/googlesitekit-wp-dashboard.js
   528951  01-27-2020 22:17   google-site-kit/dist/assets/js/vendors~chunk-googlesitekit-adminbar-9e789bf1fd190030d0dd.js
     7947  01-27-2020 22:17   google-site-kit/dist/assets/js/vendors~chunk-googlesitekit-setup-wizard-proxy-6b85c1dd02253863adf4.js
   539595  01-27-2020 22:17   google-site-kit/dist/assets/vendor/lodash.js
    73121  01-27-2020 22:17   google-site-kit/dist/assets/vendor/lodash.min.js
   146860  01-27-2020 22:17   google-site-kit/dist/assets/vendor/moment.js
    51679  01-27-2020 22:17   google-site-kit/dist/assets/vendor/moment.min.js
   726443  01-27-2020 22:17   google-site-kit/dist/assets/vendor/react-dom.development.js
   100002  01-27-2020 22:17   google-site-kit/dist/assets/vendor/react-dom.production.min.js
    98230  01-27-2020 22:17   google-site-kit/dist/assets/vendor/react.development.js
    11814  01-27-2020 22:17   google-site-kit/dist/assets/vendor/react.production.min.js
     3797  01-27-2020 22:17   google-site-kit/dist/assets/js/externals/svgxuse.js
     1342  01-27-2020 22:17   google-site-kit/dist/assets/js/externals/wp-polyfill-element-closest.js
     8878  01-27-2020 22:17   google-site-kit/dist/assets/js/externals/wp-polyfill-fetch.js
     4392  01-27-2020 22:17   google-site-kit/dist/assets/js/externals/wp-polyfill-formdata.js
     1284  01-27-2020 22:17   google-site-kit/dist/assets/js/externals/wp-polyfill-node-contains.js
     6969  01-27-2020 22:17   google-site-kit/dist/assets/js/externals/wp-polyfill-url.js
    97849  01-27-2020 22:17   google-site-kit/dist/assets/js/externals/wp-polyfill.js

```
</details>

Comment (aaemnnosttv) [reply webpack.config.js]: Also, we should update the build to remove the `copy-vendor` gulp task and other vendor/external  built files.

<details>
<summary>Diff</summary>

```diff
--- Untitled    Tue Jan 28 09:59:31 2020
+++ Untitled2   Tue Jan 28 09:59:31 2020
@@ -1,4 +1,4 @@
-BEFORE
+AFTER
      1844  01-27-2020 22:17   google-site-kit/dist/assets/js/ads.js
    973157  01-27-2020 22:17   google-site-kit/dist/assets/js/allmodules.js
     44280  01-27-2020 22:17   google-site-kit/dist/assets/js/chunk-googlesitekit-adminbar-05e75b29d0d33cbe4709.js
@@ -20,18 +20,3 @@
    589909  01-27-2020 22:17   google-site-kit/dist/assets/js/googlesitekit-wp-dashboard.js
    528951  01-27-2020 22:17   google-site-kit/dist/assets/js/vendors~chunk-googlesitekit-adminbar-9e789bf1fd190030d0dd.js
      7947  01-27-2020 22:17   google-site-kit/dist/assets/js/vendors~chunk-googlesitekit-setup-wizard-proxy-6b85c1dd02253863adf4.js
-   539595  01-27-2020 22:17   google-site-kit/dist/assets/vendor/lodash.js
-    73121  01-27-2020 22:17   google-site-kit/dist/assets/vendor/lodash.min.js
-   146860  01-27-2020 22:17   google-site-kit/dist/assets/vendor/moment.js
-    51679  01-27-2020 22:17   google-site-kit/dist/assets/vendor/moment.min.js
-   726443  01-27-2020 22:17   google-site-kit/dist/assets/vendor/react-dom.development.js
-   100002  01-27-2020 22:17   google-site-kit/dist/assets/vendor/react-dom.production.min.js
-    98230  01-27-2020 22:17   google-site-kit/dist/assets/vendor/react.development.js
-    11814  01-27-2020 22:17   google-site-kit/dist/assets/vendor/react.production.min.js
-     3797  01-27-2020 22:17   google-site-kit/dist/assets/js/externals/svgxuse.js
-     1342  01-27-2020 22:17   google-site-kit/dist/assets/js/externals/wp-polyfill-element-closest.js
-     8878  01-27-2020 22:17   google-site-kit/dist/assets/js/externals/wp-polyfill-fetch.js
-     4392  01-27-2020 22:17   google-site-kit/dist/assets/js/externals/wp-polyfill-formdata.js
-     1284  01-27-2020 22:17   google-site-kit/dist/assets/js/externals/wp-polyfill-node-contains.js
-     6969  01-27-2020 22:17   google-site-kit/dist/assets/js/externals/wp-polyfill-url.js
-    97849  01-27-2020 22:17   google-site-kit/dist/assets/js/externals/wp-polyfill.js
```

</details>

Also, maybe `ads.js` and `allmodules.js` don't need to be (or make sense to be) entrypoints anymore either? I think the only non-chunk js files we need are `googlesitekit-*.js` + e2e utilities.

File: assets/js/components/date-range-selector.js [hunk 10]
```diff
@@ -99,12 +109,13 @@ class DateRangeSelector extends Component {
 		return (
 			<Select
 				enhanced
+				se
 				className="mdc-select--minimal"
 				name="time_period"
 				label=""
 				onEnhancedChange={ this.handleSelection }
 				options={ options }
-				value={ applyFilters( this.dateRangeHook, __( 'Last 28 days', 'google-site-kit' ) ) }
```
[the remark this answers is not in the document]
Comment (felixarntz) [reply assets/js/components/date-range-selector.js]: @tofumatt Can you clarify why you're changing this here? This seems part of #447 more than this issue. If there were specific problems that this caused when rearranging the assets, would be great if you could clarify this.

Either way, I agree with @aaemnnosttv that we should either keep this around for now or eliminate the filter everywhere.

File: assets/js/googlesitekit/api/index.test.js [hunk 11]
```diff
@@ -100,7 +100,7 @@ describe( 'googlesitekit.api', () => {
 		it( 'should send query string params from data params', async () => {
 			fetch
 				.doMockIf(
-					/^\/google-site-kit\/v1\/core\/search-console\/data\/search/
+					/\/google-site-kit\/v1\/core\/search-console\/data\/search/
```
Comment (felixarntz) [inline assets/js/googlesitekit/api/index.test.js]: This is less strict than before - why did you remove the `^`?

File: assets/js/hooks-shim.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
+/**
+ * WordPress dependencies
+ */
+import { createHooks } from '@wordpress/hooks__non-shim';
+
+if ( global._GoogleSiteKitHooks === undefined ) {
+	global._GoogleSiteKitHooks = createHooks();
```
[the remark this answers is not in the document]
Comment (felixarntz) [reply assets/js/hooks-shim.js]: @tofumatt Can you explain a bit more what the problem here was? Why can't we keep this internal?

File: package.json (hunk 3 above)
Comment (felixarntz) [reply package.json]: I think it's a good practice to keep these around. We can always adjust as necessary, but it's good to have these even as a flag to us when things might get out of hand.

File: tests/e2e/mu-plugins/e2e-assets.php [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+add_action(
+	'init',
+	function () {
+		if ( ! defined( 'GOOGLESITEKIT_PLUGIN_MAIN_FILE' ) ) {
+			return;
+		}
+
+		wp_enqueue_script(
+			'googlesitekit-e2e',
```
Comment (felixarntz) [inline tests/e2e/mu-plugins/e2e-assets.php]: Let's name this `googlesitekit-e2e-utilities`.

File: assets/js/api-fetch-shim.js (hunk 8 above)
Comment (felixarntz) [reply assets/js/api-fetch-shim.js]: I'm curious as well why we need to export this as a global. The `e2e-utilities` script should do that only for the e2e tests, our production assets shouldn't expose it.

_If_ we really need to, then it should be named `_googlesitekitAPIFetch`.

File: .eslintrc.json [hunk 14]
```diff
@@ -7,18 +7,12 @@
...
   "globals": {
     "React": "readonly",
     "googlesitekit": "readonly",
     "googlesitekitAdminbar": "readonly",
     "googlesitekitDashboard": "readonly",
     "googlesitekitSettings": "readonly",
-    "lodash": "readonly",
     "googlesitekitCurrentModule": "readonly",
     "gtag": "readonly",
     "process": "readonly"
```
Comment (felixarntz) [inline .eslintrc.json]: Can't we remove a few of those that we no longer need?
* React (we shouldn't ever have this as a global)
* `googlesitekitDashboard` (doesn't exist?)
* `googlesitekitSettings` (doesn't exist?)
* `process` (what is this?)

File: assets/js/components/date-range-selector.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (tofumatt) [reply assets/js/components/date-range-selector.js]: Sure thing. I changed it to tidy up slightly but I don't think it's required here. I'll change it back 👍

[the remark this answers is not in the document]
Comment (tofumatt) [reply assets/js/components/date-range-selector.js]: (Side note: `getCurrentDateRange()` just calls `applyFilters( this.dateRangeHook, __( 'Last 28 days', 'google-site-kit' ) )`, so I just changed it here to tidy up the code, but still: I'll change it back to keep the PR a bit more atomic.

[the remark this answers is not in the document]
Comment (tofumatt) [reply assets/js/components/date-range-selector.js]: Oh, and this is just using `state` to force a component update when the date is updated, which otherwise wouldn't happen here. The values are all still the same.

File: includes/Core/Assets/Assets.php (hunk 2 above)
Comment (tofumatt) [reply includes/Core/Assets/Assets.php]: These are browser polyfills, right? I'd think Babel and webpack together are handling that… and if they aren't we should have them do it rather than separate scripts.

But some of these polyfills (eg `window.fetch`) are for aggressively out-of-date browsers 😄

These are the browsers in our `browserlist`, fwiw:
Screenshot 2020-01-28 at 16 21 54

File: package.json (hunk 3 above)
Comment (tofumatt) [reply package.json]: Agreed. We're cool with bumping them now, but it's a nice idiot check in case we accidentally include an extra MB unintentionally. 😄

File: assets/js/components/data/index.js (hunk 4 above)
Comment (tofumatt) [reply assets/js/components/data/index.js]: This was from testing, whoops! I'll remove 😄

File: webpack.config.js (hunk 5 above)
Comment (tofumatt) [reply webpack.config.js]: Yeah, I think so. Let me have a quick look into that 👍

File: .eslintrc.json (hunk 14 above)
Comment (tofumatt) [reply .eslintrc.json]: Good call. We can actually remove all of these; I've done so in a single commit (bc167dfdf47265c3bebc2ab2671235d120eadba7) in case we want to move it out of this PR, but it was easy to do and everything still works 😄

File: assets/js/googlesitekit/api/index.test.js (hunk 11 above)
Comment (tofumatt) [reply assets/js/googlesitekit/api/index.test.js]: Sorry, that was from an unrelated change/testing of something else and I forgot to restore it. Thanks for catching it, I've reverted that one! 👍

File: assets/js/components/settings/settings-modules.js [hunk 15]
```diff
@@ -211,12 +211,12 @@ class SettingsModules extends Component {
 		const { activeTab } = this.props;
 		const modulesBeingEdited = filter( isEditing, ( module ) => module );
 		const editActive = 0 < modulesBeingEdited.length;
-		if ( ! global.global.googlesitekit || ! global.global.googlesitekit.modules ) {
+		if ( ! global.googlesitekit || ! global.googlesitekit.modules ) {
```
Comment (tofumatt) [inline assets/js/components/settings/settings-modules.js]: Aha, whoops. Thanks! 😅

File: assets/js/hooks-shim.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/hooks-shim.js]: @felixarntz I don't know that we ever found out the exact cause (I think something funny happening during compilation) but basically hooks just wouldn't work at all, and raised no errors. From what I found `@wordpress/hooks` the `actions` and `filters` exported were both `null` when they were supposed to be objects [like this](https://github.com/WordPress/gutenberg/blob/4c1ef946655669bf3f031579033b75f7fae7b6d7/packages/hooks/src/createHooks.js#L18-L21).

I took a look at this again yesterday to investigate and found that hooks are mostly working now without this global but some elements of the UI are missing so it might be worth digging into this a little more. This may have changed due to other changes/fixes in this branch that seemed unrelated (such as opting out of amd parsing), but it's hard to say.

File: assets/js/api-fetch-shim.js [hunk 16]
```diff
@@ -3,9 +3,7 @@
  */
 import apiFetch from '@wordpress/api-fetch__non-shim';
 
-if ( global._apiFetch === undefined ) {
-	global._apiFetch = apiFetch;
-
+if ( global._apiFetchRootURL || global._apiFetchNonceMiddleware ) {
```
Comment (aaemnnosttv) [inline assets/js/api-fetch-shim.js]: Can we not source these from `global._googlesitekitBase` (would need to be added of course) rather than generic globals?

Comment (tofumatt) [reply assets/js/api-fetch-shim.js]: Indeed! I just haven't got around to it yet. I'll do that ASAP 👍

**Comment by tofumatt** [comment]
*Whew!*

I think this is finally ready for an exhaustive PR. I added a quick list of changes made to support this PR; I know it's way bigger than we thought it'd be, but we kept unearthing a lot of papercuts and bad assumptions the code was making, frequently around global/shared state with WordPress. It's good to fix these up as—while a slow-to-change API—WordPress could have changed JS things on us and broken our builds in production.

This should prevent that but also provides us with other wins, outlined in the PR summary as well.

Apologies for the massive diff, and thanks a lot to @aaemnnosttv for all his help on this PR. The code base should be a lot more explicit now! 🎉

File: includes/Core/Assets/Assets.php [hunk 17]
```diff
@@ -373,12 +366,17 @@ private function get_assets() {
...
 					'before_print' => function( $handle ) {
 						wp_add_inline_script(
 							$handle,
-							'window._googlesitekitBase = ' . wp_json_encode( $this->get_inline_base_data() ),
+							implode(
+								"\n",
+								array(
+									'window._googlesitekitBase = ' . wp_json_encode( $this->get_inline_base_data() ),
+								)
+							),
```
Comment (felixarntz) [inline includes/Core/Assets/Assets.php]: Why this change? That looks a bit unnecessary to me :)

File: includes/Core/Assets/Assets.php [hunk 18]
```diff
@@ -513,13 +508,16 @@ private function get_inline_base_data() {
...
-			'homeURL'          => home_url(),
-			'referenceSiteURL' => esc_url_raw( $site_url ),
-			'userIDHash'       => md5( $site_url . $current_user->ID ),
-			'adminRoot'        => esc_url_raw( get_admin_url() . 'admin.php' ),
-			'assetsRoot'       => esc_url_raw( $this->context->url( 'dist/assets/' ) ),
-			'blogPrefix'       => $wpdb->get_blog_prefix(),
-			'isNetworkMode'    => $this->context->is_network_mode(),
+			'apiFetchNonceEndpoint'   => admin_url( 'admin-ajax.php?action=rest-nonce' ),
+			'apiFetchNonceMiddleware' => ( wp_installing() && ! is_multisite() ) ? '' : wp_create_nonce( 'wp_rest' ),
+			'apiFetchRootURL'         => esc_url_raw( get_rest_url() ),
```
Comment (felixarntz) [inline includes/Core/Assets/Assets.php]: These should not be in here. One goal of moving away from the `googlesitekit` global for passing PHP data to JS is to split responsibilities between multiple globals, not have one as a dump for any data.

* The `_googlesitekitBase` one should only be used for base data about the environment of the current site or current user, which is needed across multiple places, and even in the most basic "standalone" scripts. A rough guidance should be that this is data we would potentially also expose via the PHP `Context` class.
* All these `apiFetch...` values are very specific to the `apiFetch` function, already indicated by the prefix. If anything, the REST URL itself can be considered base data (in which case it should be named `restRoot` though, like the other ones below. However, I think for now we should put all of these in a separate `_googlesitekitAPIFetchData` global, which should only be used by our `apiFetch` function.

File: assets/js/modules/analytics/dashboard/dashboard-widget-popular-pages-table.js [hunk 19]
```diff
@@ -61,7 +61,7 @@ class AnalyticsDashboardWidgetPopularPagesTable extends Component {
 
 	render() {
 		const { data } = this.props;
-		const { siteURL: siteURL } = googlesitekit.admin;
+		const { siteURL: siteURL } = global.googlesitekit.admin;
```
Comment (aaemnnosttv) [inline assets/js/modules/analytics/dashboard/dashboard-widget-popular-pages-table.js]:
```suggestion
		const { siteURL } = global.googlesitekit.admin;
```

File: tests/phpunit/integration/Core/Admin/ScreenTest.php [hunk 20]
```diff
@@ -103,7 +103,6 @@ public function test_initialize() {
 
 	public function test_enqueue_assets() {
 		wp_dequeue_style( 'googlesitekit_admin_css' );
-		wp_dequeue_script( 'googlesitekit_modules' );
```
Comment (felixarntz) [inline tests/phpunit/integration/Core/Admin/ScreenTest.php]: Instead of removing this, let's exchange it with another script that we still have, to make sure enqueuing scripts is still covered.

File: package.json [hunk 21]
```diff
@@ -165,11 +165,14 @@
...
     "jest-localstorage-mock": "^2.4.0",
+    "jquery": "^3.4.1",
     "lint-staged": "^8.1.5",
-    "lodash": "^4.17.11",
+    "lodash": "^4.17.15",
+    "lodash-es": "^4.17.15",
     "lodash-webpack-plugin": "^0.11.5",
     "md5": "^2.2.1",
     "mini-css-extract-plugin": "^0.4.5",
+    "moment": "^2.24.0",
```
Comment (felixarntz) [inline package.json]: Why was this added here? Are we actually using `moment` anywhere? I see it was in our code before, but I'm wondering why.

File: .storybook/webpack.config.js [hunk 22]
```diff
@@ -1,4 +1,5 @@
 const path = require( 'path' );
+const MiniCssExtractPlugin = require( 'mini-css-extract-plugin' );
 
 module.exports = async ( { config } ) => {
 	config.resolve = {
```
Comment (aaemnnosttv) [inline .storybook/webpack.config.js]: @tofumatt can we update this to resolve `@wordpress/api-fetch` with an alternate version just for Storybook which returns mocked data since the below will no longer work?

https://github.com/google/site-kit-wp/blob/7c192a71c09610e95d2d5fbcefad16b7abbe9d71/.storybook/config.js#L89-L111

File: assets/js/components/date-range-selector.js [hunk 23]
```diff
@@ -85,10 +86,17 @@ class DateRangeSelector extends Component {
 		// Trigger a data refresh.
 		doAction( 'googlesitekit.moduleDataReset' );
 		doAction( 'googlesitekit.moduleLoaded', context );
+
+		// Update this component.
+		this.setState( {
+			dateValue: applyFilters( this.dateRangeHook, __( 'Last 28 days', 'google-site-kit' ) ),
+		} );
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/components/date-range-selector.js]: @tofumatt  can correct me but I believe it was necessary due to an issue with the component not re-rendering.

**Comment by felixarntz** [comment]
@aaemnnosttv @tofumatt Just tested how these changes affect our bundled ZIP, the file size is now up to 3.4 MB from about 2.9 MB. While that is not insignificant, I think it's okay given the nature of these changes.

File: tests/phpunit/integration/Core/Admin/ScreenTest.php (hunk 20 above)
Comment (aaemnnosttv) [reply tests/phpunit/integration/Core/Admin/ScreenTest.php]: I did this in [`AssetsTest`](https://github.com/google/site-kit-wp/pull/1013/files#diff-3dff40ce7dc41bb9397872c63bf36a96R102) since this test is more specific to assets enqueued globally across many screens. Of course adding back a different asset would not be a problem though. Apart from fonts, `googlesitekit_admin_css` is the only asset enqueued on all screens.
https://github.com/google/site-kit-wp/blob/b37b4abf8afb11f61b3727b5206064b61ab9ee4e/includes/Core/Admin/Screen.php#L177-L192

File: includes/Core/Assets/Assets.php (hunk 17 above)
Comment (aaemnnosttv) [reply includes/Core/Assets/Assets.php]: Ah, I think this was combined with a few others at some point that were later (re)moved. I'll revert this portion 👍

Comment (aaemnnosttv) [reply includes/Core/Assets/Assets.php]: Fixed in b7b8cad596c2d2409bd27a29f6fb3a99a0b6a2b2

File: includes/Core/Assets/Assets.php (hunk 18 above)
Comment (aaemnnosttv) [reply includes/Core/Assets/Assets.php]: Sounds good @felixarntz. Seeing as we need the apiFetch data in all the same places, do you think it would be OK then to define this alternate global before or after the existing definition of `window._googlesitekitBase`?

Comment (felixarntz) [reply includes/Core/Assets/Assets.php]: @aaemnnosttv We should not be needing `apiFetch` in `googlesitekit-admin.js` for example, so I don't think the two objects are generally used in the same places.

For now I'm okay including them in the same place, they should be split out however via #1004.

File: tests/phpunit/integration/Core/Admin/ScreenTest.php (hunk 20 above)
Comment (felixarntz) [reply tests/phpunit/integration/Core/Admin/ScreenTest.php]: SGTM

File: package.json (hunk 21 above)
Comment (tofumatt) [reply package.json]: Odd, I don't see it in use anywhere in our codebase. I'd say it's safe to remove entirely.

**Comment by aaemnnosttv** [comment]
@felixarntz what do think about dropping the E2E test for 4.9+Gutenberg now that we don't share any dependencies with it?

File: includes/Core/Assets/Assets.php (hunk 18 above)
Comment (aaemnnosttv) [reply includes/Core/Assets/Assets.php]: Updated!

**Comment by felixarntz** [comment]
@aaemnnosttv 
> what do think about dropping the E2E test for 4.9+Gutenberg now that we don't share any dependencies with it?

Let's keep these for now, we'll still want to make sure everything works in that combination. It will be less likely to have specific breakage there, but I'd prefer to remove this only a few cycles later.

**Comment by tofumatt** [comment]
That increase in bundle size is in-line with what I expected—as mentioned, it's pretty good all-things-considered and I think worth the tradeoff.

I've tweaked the Storybook `apiFetch` handling so this is all set from me—I think everything from @felixarntz and @aaemnnosttv has been addressed so it's set for another look. 👍
